compiler-plugin/src/main/kotlin/arrow/meta/autofold/AutoFoldPlugin.kt

```diff
@@ -0,0 +1,102 @@
+package arrow.meta.autofold
+
+import arrow.meta.extensions.ExtensionPhase
+import arrow.meta.extensions.MetaComponentRegistrar
+import arrow.meta.higherkind.arity
+import arrow.meta.higherkind.invariant
+import arrow.meta.qq.classOrObject
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.name.Name
+import org.jetbrains.kotlin.psi.KtClass
+import org.jetbrains.kotlin.psi.KtClassOrObject
+import org.jetbrains.kotlin.psi.KtDeclaration
+import org.jetbrains.kotlin.psi.psiUtil.findFunctionByName
+import org.jetbrains.kotlin.psi.psiUtil.getSuperNames
+
+val MetaComponentRegistrar.autoFold: List<ExtensionPhase>
+  get() =
+    meta(
+      classOrObject(::isAutoFoldable) { c ->
+        println("Processing Sealed class: ${c.name}")
+        sealedVariants.forEach(::println)
+        val sealedExtraTypes = sealedVariants.map { it.typeVariables }.filter(String::isNotEmpty)
+        val typeInfo = (c.renderTypeParameters().split(", ") + sealedExtraTypes).distinct().sorted()
```

Comment:
there is gotta be a better way to do string manipulation here. What does `typeinfo` refer to?